components/collections/src/codepointtrie/cptrie.rs

```diff
@@ -485,43 +507,56 @@ impl<'trie, T: TrieValue> CodePointTrie<'trie, T> {
     /// Returns the value that is associated with `code_point` in this [`CodePointTrie`]
     /// assuming that the fast index path should be used.
     ///
-    /// # Safety
+    /// # Intended Precondition
     ///
     /// `code_point` must be at most `FAST_TYPE_FAST_INDEXING_MAX` if
     /// the trie type is fast or at most `SMALL_TYPE_FAST_INDEXING_MAX`
     /// if the trie type is small.
     ///
+    /// To avoid having to make this method `unsafe`, the above condition
+    /// is checked in the hope that the optimizer deletes the check due
+    /// to the caller already having checked it correctly. However, if
+    /// the caller is incorrect, a violation of this precondition results
+    /// in a panic when debug assertions are enabled and in returning
+    /// `self.error_value` when they aren't.
+    ///
     /// # Panics
     ///
     /// When debug assertions are enabled, the above precondition and
     /// slice access bounds are checked.
-    #[inline(always)]
-    unsafe fn get32_by_fast_index(&self, code_point: u32) -> T {
-        debug_assert!(
-            code_point
-                <= match self.header.trie_type {
-                    TrieType::Fast => FAST_TYPE_FAST_INDEXING_MAX,
-                    TrieType::Small => SMALL_TYPE_FAST_INDEXING_MAX,
-                }
-        );
+    #[inline(always)] // "always" to make the `fast_max` check collapse away.
```

Comment:
nit: Oh. I was suggesting this be written as `get32_by_fast_index() -> Option<T>`

CSE optimizations where you hope that a duplicated bounds check is collapsed are a bit less reliable than just collapsing an Option check into a single branch.

Then you won't need the linker hack, and it'll optimize better.

Reply:
I see. That makes more sense. Changed. Thanks.